R/utils.R

```diff
@@ -116,3 +133,88 @@ recycle <- function(x, n, name = deparse(substitute(x))) {
     stop(sprintf("Invalid length for '%s', expected 1 or %d", name, n))
   }
 }
+
+
+is_3d_array <- function(x) {
+  !is.na(nlayer(x))
+}
+
+
+## Copied from ncol
+nlayer <- function(x) {
+  dim(x)[[3L]]
+}
+
+
+## Copied from NCOL
+NLAYER <- function(x) { # nolint
+  d <- dim(x)
+  if (length(d) > 2L) {
+    d[3L]
+  } else {
+    1L
+  }
+}
+
+
+## Copied from colnames
+layernames <- function(x, do.NULL = TRUE, prefix = "layer") { # nolint
+    dn <- dimnames(x)
+    if (!is.null(dn[[3L]]))
+        dn[[3L]]
+    else {
+      if (do.NULL) {
+        NULL
+      } else {
+        paste0(prefix, seq_len(NLAYER(x)))
+      }
+    }
+}
+
+
+## Copied from colnames<-
```

Comment:
rather than a copy (with copyright issues) can you adapt this to the subset that we need perhaps?